### PR TITLE
Two Factor Authentication

### PR DESCRIPTION
**Pull request by ricknjacky**
- [x] Enable/Disable TOTP
- [x] Verification of TOTP entered
- [x] Login Verification
- [x] Unit Tests
- [x] Documentation

### CODE REVIEW

**Comment by llaske** [comment]
I can't see the new Enable 2FA button on the user profile.
Is there something special to configure?

image

**Comment by ricknjacky** [comment]
Right, the settings is in the profile view (the one we access from top right corner menu) I rechecked, with this branch's file changes, It works.

^^Do test this at your end too and let me know if any issues occur here.

**Comment by llaske** [comment]
> we are restricting access to making changes to 2FA's state and enable/disable to the user only (account holder)

Make sense.

Few remarks:

* The popup "do you really want to activate 2fa settings?" is useless since it's possible to go back without change.

Capture d’écran du 2021-10-05 21-52-36

* When 2fa settings is enable the button label change to "disable 2fa authentication". I don't think you need to change the label.

Capture d’écran du 2021-10-05 21-55-39

* Could you rename `SugarizerServer` to `Sugarizer Server` (white space between words) in the 2FA name? Is it possible to add the Sugarizer icon?

image

* When 2fa settings is enable. I expected to be asked a 2FA when login but nothing happen even. Do I miss something?

Peek 05-10-2021 22-17

**Comment by llaske** [comment]
> Should I create a view for twoFactor.ejs? I mean to say that there is nothing to show really when 2Factor is enabled. So we decided, it's better to have the disable button on profile view itself. Let me know if I should move this logic to twoFactor.ejs page.

Yes. Because it's not possible to change 2FA settings from the profile page, I think it's better to put everything related to 2FA in a dedicated page.

> Yes you can, it's up to the user whatever name they wish to give it. You can change it from here

Nice. Good to hear. Is it possible easily to add a Sugarizer icon?

> Can you please tell me which commit you're on?

I'm using the latest commit in this branch.

**Comment by ricknjacky** [comment]
> > Should I create a view for twoFactor.ejs? I mean to say that there is nothing to show really when 2Factor is enabled. So we decided, it's better to have the disable button on profile view itself. Let me know if I should move this logic to twoFactor.ejs page.
> 
> Yes. Because it's not possible to change 2FA settings from the profile page, I think it's better to put everything related to 2FA in a dedicated page.

Got it, will integrate the feedback.

> > Yes you can, it's up to the user whatever name they wish to give it. You can change it from here
> 
> Nice. Good to hear. Is it possible easily to add a Sugarizer icon? 

I don't think so, because most authenticator apps don't have a profile picture kind of a thing. For instance,  here's a screenshot of Google authenticator's UI:- [here](https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcT1vx6AovU3TRCXPpZKl7vCDOVY3or68gPN8w&usqp=CAU)


> 
> > Can you please tell me which commit you're on?
> 
> I'm using the latest commit in this branch.

Weird, I'll have to inspect.

**Comment by llaske** [comment]
@ricknjacky I've found a regression on your PR. When I try to connect to a Sugarizer user it failed.
See below.

Peek 10-10-2021 16-10

**Comment by llaske** [comment]
Nice @ricknjacky. It works now.

Few remarks:

- Buttons alignment is not good. I think it should be `margin-top: -63px` for Enable button and `margin-top: -68px` for Disable button.
- Put 2FA tutorial popup before the created time popup. Popup should appear in the field order.
- Remove commented code [here](https://github.com/llaske/sugarizer-server/pull/285/files#diff-bb2a36ef01b2061d21767920986a5cfe9825896503ccf12f051e3fee1fc2de13R12-R13).
- Add a white space between Sugarizer and Server on the default service name [here](https://github.com/llaske/sugarizer-server/pull/285/files#diff-2af86d17090d1ed1caa8cd099561ebd56e029713c325d237bb02dff38ef3275fR18).

File: api/controller/users.js [hunk 1]
```diff
@@ -101,6 +104,218 @@ exports.findById = function(req, res) {
...
+						} else if (result) {
+							var updatedUser = result.value;
+							var name = updatedUser.name;
+							var manualKey = updatedUser.uniqueSecret;
+							var otpAuth = generateOTPToken(name, serviceName, manualKey);
+
+							delete updatedUser.password;
+							//send user, manualKey and otpAuth for QRCode async function.
+							res.send({
+								user: user,
```
Comment (NikhilM98) [inline api/controller/users.js:161]: @ricknjacky return updatedUser here

File: api/controller/users.js [hunk 2]
```diff
@@ -101,6 +104,218 @@ exports.findById = function(req, res) {
...
+
+	//disable TOTP
+	db.collection(usersCollection, function(err, collection) {
+		collection.findOneAndUpdate({
+			'_id': new mongo.ObjectID(uid)
+		}, {
+			$set:
+				{
+					tfa: false,
+					uniqueSecret: null
```
Comment (NikhilM98) [inline api/controller/users.js:291]: Replace null with undefined

File: api/controller/auth.js [hunk 3]
```diff
@@ -122,6 +131,64 @@ exports.login = function(req, res) {
...
+			var maxAgeTfa = req.iniconfig.security.max_age_TFA;
+
+			if (isValid === true) {
+				delete user.uniqueSecret;
+				// refresh the user token and set partial to false.
+				res.send(genToken(user, maxAge, false));
+			} else {
+				delete user.deployments;
+				delete user.uniqueSecret;
+				console.log(user);
```
Comment (NikhilM98) [inline api/controller/auth.js:180]: Remove unnecessary console.log statements.

File: api/controller/auth.js [hunk 4]
```diff
@@ -122,6 +131,64 @@ exports.login = function(req, res) {
...
+	}, {enableSecret: true}, function(users) {
+		if (users && users.length > 0) {
+
+			//take the first user incase of multple matches
+			var user = users[0];
+			var uniqueSecret = user.uniqueSecret;
+			try {
+				var isValid = otplib.authenticator.check(uniqueToken, uniqueSecret);
+			} catch (err) {
+				console.log(err.message);
```
Comment (NikhilM98) [inline api/controller/auth.js:163]: Remove unnecessary console.log statements.

File: api/controller/utils/qrCodeUtil.js [hunk 5]
```diff
@@ -0,0 +1,13 @@
+var qrcode = require('qrcode');
+
+exports.generateQRCode = async function (otpAuth) {
+
+    try {
+        var QRCodeImageUrl = await qrcode.toDataURL(otpAuth);
+        return { image: `<img src='${QRCodeImageUrl}' alt='qr-code-img' />` };
+    } catch (error) {
+        console.log('Could not generate QR code', error);
+        return;
```
Comment (NikhilM98) [inline api/controller/utils/qrCodeUtil.js:10]: Either return html instead of undefined or handle undefined.

File: api/controller/auth.js [hunk 6]
```diff
@@ -122,6 +131,64 @@ exports.login = function(req, res) {
...
+
+			if (isValid === true) {
+				delete user.uniqueSecret;
+				// refresh the user token and set partial to false.
+				res.send(genToken(user, maxAge, false));
+			} else {
+				delete user.deployments;
+				delete user.uniqueSecret;
+				console.log(user);
+				res.send(genToken(user, maxAgeTfa, true)); // refresh the token -- todo limit the number of attempts.
```
Comment (NikhilM98) [inline api/controller/auth.js:181]: Todo comment is not needed

File: api/controller/auth.js [hunk 7]
```diff
@@ -307,16 +374,18 @@ exports.checkAdminOrLocal = function(req, res, next) {
...
 
 // private method
-function genToken(user, age) {
+function genToken(user, age, partial) {
 	var expires = expiresIn(age);
 	var token = jwt.encode({
+		partial: partial,
 		exp: expires
-	}, require('../../config/secret')());
+	}, secret);
```
Comment (NikhilM98) [inline api/controller/auth.js:382]: Great!

File: package.json [hunk 8]
```diff
@@ -52,4 +54,4 @@
 		"seed:users-csv": "node scripts/seed_users.js",
 		"lint": "eslint .; exit 0"
 	}
-}
+}
```
Comment (NikhilM98) [inline package.json:57]: Add a new line here

**Comment by llaske** [comment]
Thanks @ricknjacky for fixing PR feedback.

Two things left to do : 

- [ ] Add a warning message when a teacher enable 2FA 
- [ ] Update unit tests due to 2FA changes

**Comment by ricknjacky** [comment]
> Thanks @ricknjacky for fixing PR feedback.
> 
> Two things left to do :
> 
> * [ ]  Add a warning message when a teacher enable 2FA
> * [ ]  Update unit tests due to 2FA changes

Thank-you for the update!
Adding warning message asap.
As discussed, once this PR is merged into dev, I'll rebase and send in a new PR for tests.

**Comment by llaske** [comment]
Regarding unit test if it's not too complex it will be better to report your changes here so everything will be at the same place

**Comment by ricknjacky** [comment]
@llaske 
Screenshot from 2021-11-13 07-22-55
Have added all the relevant test cases and they all pass successfully. Kindly review, Thanks.

**Comment by llaske** [comment]
Nice @ricknjacky.
Could you also update the Server Settings in README file to explain new parameters meaning in .INI file?

**Comment by ricknjacky** [comment]
@llaske done. Added relevant comments and information via comments too.

**Comment by llaske** [comment]
Good job @ricknjacky . Thanks for this contribution.

**Pull request merged**